dmcontent/questions.py

```diff
@@ -480,6 +479,51 @@ def summary(self, service_data):
         return ListSummary(self, service_data)
 
 
+class Hierarchy(List):
+    """
+    For our purposes, a Hierarchy is like a List, except the entries
+    are potentially related to each other in a "subsumptive
+    containment hierarchy". For example, every service that is a member
+    of a child category is _by definition_ also a member of its parent
+    category as well, and we can add those parent categories as a
+    denormalization to make searching more efficient.
+    """
+
+    def _get_data(self, form_data):
+        if self.id not in form_data:
+            return {self.id: None}
+
+        values = set(form_data.getlist(self.id))
+        values.update(self.get_missing_values(values))
+
+        return {self.id: sorted(values) or None}
+
+    def summary(self, service_data):
+        return HierarchySummary(self, service_data)
+
+    def get_missing_values(self, selected_values_set):
+        """
+        Recursively retrieves un-selected parent categories of the
+        passed-in selection, as a set of 'value' strings.
+        :param selected_values_set: initially-selected categories (e.g. by the user)
+        :return: additional values that should also be selected
+        """
+
+        def update_expected_values(options, parents, expected_set):
```

Comment:
Ah, I see your point. Since this is not a method, returning a value feels more natural, but you're right in that you'd have to copy to avoid mutation.

I think in this case it'd be ok to pass in `set()` as an argument directly and then there's no side-effect since we don't have a reference. You could also default `expected_values` to `None` and then replace `None` with `set()` in the function itself, avoiding the need to pass it in as the argument.

But that's probably only marginally better, so I'm happy with keeping this as is.